### ISSUE TITLE
Allow specifying maximum number of connections on Server

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#1886]
This will probably be implemented as part of a larger piece of work to control resources on server-side.

**Comment by asif31iqbal** [comment]
Thanks for creating this issue.

**Comment by rmichela** [comment]
What's the status of gRPC resource constraints?

**Comment by abaldeva** [comment]
+1 for this.

**Comment by hehe717** [comment]
any updates for this issues?

**Comment by ejona86** [comment]
No updates. Such a feature should be cross-language, and I confirmed that with the other languages today.  I guess it would be helpful to know the more precise use cases (I know use cases for this, but I want to know your reason for asking) and expected behavior when the limit is exceeded. We would also need to decide how "big" of a feature this becomes, since I do think there are various levels of completeness and their associated complexity that the solutions could have.

**Comment by Daniel-B-Smith** [comment]
I can comment about why we would like it at Datadog for Golang. We're actively working on throttlers to prevent OOMs when a server becomes overloaded. We can go a long way via throttling inside our application, but there is still the risk that memory usage at the OS/connection layer will push us over the limit. Ideally in my opinion, when the server hits too many connections, failed connection attempts will return RESOURCE_EXHAUSTED to the client.

**Comment by ejona86** [comment]
> Ideally in my opinion, when the server hits too many connections, failed connection attempts will return RESOURCE_EXHAUSTED to the client.

This is not feasible. Failed connections cause UNAVAILABLE on the client. To cause RESOURCE_EXHAUSTED we'd have to accept the connection and respond with RESOURCE_EXHAUSTED for each inbound RPC.

**Comment by Daniel-B-Smith** [comment]
Yeah, that makes sense. Are there separate timeouts for connection establishment and the full request? I would rather establish the connection and fail as early as possible than sit around and timeout.

**Comment by ejona86** [comment]
@Daniel-B-Smith wrote:
> Are there separate timeouts for connection establishment and the full request?

Yes. Connection timeouts and RPC timeouts are separate. An RPC will wait if any connection attempts are in progress, but if the connections are in a known-bad state then the RPC can fail immediately. You can get a rough view with https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md . Roughly speaking (load balancing makes things complicated), RPCs will fail if the Channel is in TRANSIENT_FAILURE.

https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md is for connection establishment. It is _supposed_ to start with a 20 second timeout, and in extreme cases increase as the backoff increases over 20 seconds. But I know Java isn't implementing that; it's more of hard-coded timeouts.

**Comment by joseret** [comment]
Any news on this?  or workaround?

**Comment by khsahaji** [comment]
Facing the same issue with GoLang. Context is getting cancelled due to max number of parallel connections being hit at gRpc stream server. I need to get that error in client side and handle it but the client right now uses "SendMsg" function and finishes sending message before server is hitting the snag. All this leads to lost messages. How to handle this now ? should the client wait for the number of active connection to go down ? How does it get the number of active parallel connection to gRpc stream server ?

**Comment by ejona86** [comment]
@khsahaji, open an issue on the grpc-go repo. If you think it helps, you can reference this issue from the issue you create.

**Comment by SaxyPandaBear** [comment]
Is there any traction on making this feature a reality? I think my current options are:
1. Just allow an unbounded number of connections until the gRPC server under the hood can't serve any more concurrent channels (apologies if I flub any terminology, still an infant with gRPC)
2. Create a server interceptor that uses a global semaphore to track the number of requests and start rejecting things when the semaphore is completely used up

**Comment by angoenka** [comment]
+1
At Beam we have a use case where we use grpc with python as client and java as server in stream - stream fashion.
Each stream requires a reservation of 64MB and release the resources once the scream is closed.
This is causing OOM as a single client can create multiple connections in a burt.
We want to limit the concurrent streams to limit the resource. 
We have a PR out which delays the allocation of resources using semaphore here.
https://github.com/apache/beam/pull/9647 
Can we use a build in way in GRPC to do this?

**Comment by ejona86** [comment]
@angoenka, for a single client bursting, you can set [maxConcurrentCallsPerConnection](https://grpc.github.io/grpc-java/javadoc/io/grpc/netty/NettyServerBuilder.html#maxConcurrentCallsPerConnection-int-). Granted, that isn't dynamic so you can't _allow_ a single client to burst higher if other clients aren't using resources.

Your specific case may also not work as well for this issue by itself, since you probably want to limit that one _method_ differently from other methods (not limit for the entire server).

**Comment by rmichela** [comment]
Netflix’s concurrency limits interceptor is an interesting solution to managing server concurrency. https://github.com/Netflix/concurrency-limits

**Comment by bruno-kakele** [comment]
I am also interested in this. For instance I have a server that accepts streaming rpcs, but, a single user could spawn multiple connections and kill one of my tasks (due to excessive resource usage and also too many active connections, even ones that don't have an active stream). I am interested in how to guard a rpc server against client abuse, either a malicious client or just because there is a bug. Are there any best practices/recommendations/docs about this? Thanks in advance

**Comment by macmv** [comment]
Also interested in this. I have a server getting a stream of data, but that would be completely broken if two clients would connect. I don't have any way to limit the number of connections, which makes this difficult.

**Comment by ejona86** [comment]
@macmv, for application-level limits like that, I think you should have an interceptor that limits the number of concurrent RPCs/clients to one particular service or method. Port-level connection limits are too heavy-handed for your use-case and would not work if you have multiple services, which includes things like health checking, stats retrieval, and debugging services.

**Comment by fengylm** [comment]
Any news on this? or workaround?
